src/meepgeom.cpp

```diff
@@ -1144,6 +1144,75 @@ void geom_epsilon::eff_chi1inv_matrix(meep::component c, symmetric_matrix *chi1i
 
 static int eps_ever_negative = 0;
 static meep::field_type func_ft = meep::E_stuff;
+static double voxel_rad = 0;
+static vector3 voxel_cen;
+
+#ifdef CTL_HAS_COMPLEX_INTEGRATION
+static cnumber matgrid_ceps_func(int n, number *x, void *geomeps_) {
+  geom_epsilon *geomeps = (geom_epsilon *)geomeps_;
+  vector3 p = {0, 0, 0};
+  p.x = x[0];
+  p.y = n > 1 ? x[1] : 0;
+  p.z = n > 2 ? x[2] : 0;
+  double s = 1;
+  if (dim == meep::Dcyl) {
+    double py = p.y;
+    p.y = p.z;
+    p.z = py;
+    s = p.x;
+  }
+  cnumber ret;
+  double r2 = voxel_rad*voxel_rad;
+  double z2 = (p.x-voxel_cen.x)*(p.x-voxel_cen.x) + (p.y-voxel_cen.y)*(p.y-voxel_cen.y);
+  double w = 2*sqrt(r2 - z2)/(meep::pi*r2);
+  double ep = geomeps->chi1p1(func_ft, vector3_to_vec(p));
```

Comment:
You shouldn't be calling this.  Given u and du/dz = ‖∇u‖ at the voxel center, you don't need any further query of the geometry tree or the material grid.  You just take the first-order Taylor approximation for u(z), project it, and then compute ε or 1/ε (for the two averages, assuming the simple case of scalar ε for now).

Reply:
Fixed.